overlord/snapstate/snapstate.go

```diff
@@ -1114,6 +1139,79 @@ func RevertToRevision(st *state.State, name string, rev snap.Revision, flags Fla
 	return doInstall(st, &snapst, snapsup)
 }
 
+// TransitionCore transitions from an old core snap name to a new core
+// snap name. It is used for the ubuntu-core -> core transition (that
+// is not just a rename because the two snaps have different snapIDs)
+//
+// Note that this function makes some assumptions like:
+// - no aliases setup for both snaps
+// - no data needs to be copied
+// - all interfaces are absolutely identical on both new and old
+// Do not use this as a general way to transition from snap A to snap B.
+func TransitionCore(st *state.State, oldName, newName string) ([]*state.TaskSet, error) {
+	var oldSnapst, newSnapst SnapState
+	err := Get(st, oldName, &oldSnapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	if !oldSnapst.HasCurrent() {
+		return nil, fmt.Errorf("cannot transition snap %q: not installed", oldName)
+	}
+
+	// FIXME: what user ID to use for an automatic transition?
+	userID := 0
+	newInfo, err := snapInfo(st, newName, oldSnapst.Channel, snap.R(0), userID)
+	if err != nil {
+		return nil, err
+	}
+
+	var all []*state.TaskSet
+	// install new core (if not already installed)
+	err = Get(st, newName, &newSnapst)
+	if err != nil && err != state.ErrNoState {
+		return nil, err
+	}
+	if !newSnapst.HasCurrent() {
+		// start by instaling the new snap
+		tsInst, err := doInstall(st, &newSnapst, &SnapSetup{
+			Channel:      oldSnapst.Channel,
+			DownloadInfo: &newInfo.DownloadInfo,
+			SideInfo:     &newInfo.SideInfo,
+		})
+		if err != nil {
+			return nil, err
+		}
+		all = append(all, tsInst)
+	}
+
+	// then transition the interface connections over
+	transIf := st.NewTask("transition-connections-core-migration", fmt.Sprintf(i18n.G("Transition security profiles from %q to %q"), oldName, newName))
+	transIf.Set("old-name", oldName)
+	transIf.Set("new-name", newName)
+	if len(all) > 0 {
+		transIf.WaitAll(all[0])
+	}
+	tsTrans := state.NewTaskSet(transIf)
+	all = append(all, tsTrans)
+
+	// FIXME: this is just here for the tests
+	transIf.Set("snap-setup", &SnapSetup{
+		SideInfo: &snap.SideInfo{
+			RealName: oldName,
+		},
+	})
+
+	// then remove the old snap
+	tsRm, err := Remove(st, oldName, snap.R(0))
+	if err != nil {
+		return nil, err
+	}
+	tsRm.WaitFor(transIf)
```

Comment:
Unless I miss something the branch is doing what you describe. If the new core is already installed it won't get installed again and the old core gets removed. Once the old core is gone it can never be installed again, we check for this in checkCoreName(). But it sounds like I may have overlooked something, so I'm keen to learn more!